jgrapht-core/src/main/java/org/jgrapht/Graphs.java

```diff
@@ -433,6 +435,228 @@
         }
         return list;
     }
+
+    /**
+     * This function is implemented recursively. Therefore only graphs without
+     * any cycles are allowed.
+     *
+     * @param graph graph to look for ancestors in.
+     * @param vertex the vertex to get the ancestors of.
+     *
+     * @return {@link List} of ancestors of the vertex in the given graph.
+     */
+    public static <V, E extends DefaultEdge> List<V> determineAncestors(
+        DirectedAcyclicGraph<V, E> graph,
+        V vertex)
+    {
+
+        List<V> ancestors = new ArrayList<>();
+        ancestors = determineAncestors(graph, vertex, ancestors);
+        return ancestors;
+    }
+
+    protected static <V, E extends DefaultEdge> List<V> determineAncestors(
+        DirectedAcyclicGraph<V, E> graph,
+        V vertex,
+        List<V> ancestors)
+    {
+
+        List<V> parents = Graphs.predecessorListOf(graph, vertex);
+
+        for (V parent : parents) {
+            ancestors.add(parent);
+            determineAncestors(graph, parent, ancestors);
+        }
+
+        return ancestors;
+    }
+
+    /**
+     * This function is implemented recursively. Therefore only graphs without
+     * any cycles are allowed.
+     *
+     * @param graph graph to look for descendants in.
+     * @param vertex the vertex to get the descendants of.
+     *
+     * @return {@link List} of descendants of the vertex in the given graph.
+     */
+    public static <V, E extends DefaultEdge> List<V> determineDescendants(
+        DirectedAcyclicGraph<V, E> graph,
+        V vertex)
+    {
+
+        List<V> descendants = new ArrayList<>();
+        descendants = determineDescendants(graph, vertex, descendants);
+        return descendants;
+    }
+
+    protected static <V, E extends DefaultEdge> List<V> determineDescendants(
+        DirectedAcyclicGraph<V, E> graph,
+        V vertex,
+        List<V> descendants)
+    {
+
+        List<V> children = Graphs.successorListOf(graph, vertex);
+
+        for (V child : children) {
+            descendants.add(child);
+            determineDescendants(graph, child, descendants);
+        }
+
+        return descendants;
+    }
+
+    /**
+     * Removes the given vertex from the given graph. If the vertex to be
```

Comment:
What about the following terminology?

- successors
- direct successors
- predecessors
- direct predecessors

Therefore e.g.:

- `determineDescendants` -> `determineSuccessors`
- `removeVertexAndConnectParentsWithChildren` -> `removeVertexAndConnectDirectPredecessorsWithDirectSuccessors`
- `vertexHasChildren` -> `vertexHasDirectSuccessors`